### PR TITLE
MQTT: add auth and topic filtering

### PR DESCRIPTION
**Pull request by kruton**
Petnet devices will get a unique API key with which they will log into
the MQTT broker. The Petnet devices will also only be able to subscribe
to the topics of their own gateway IDs.

The server has its own login that is generated whenever the server
starts up. That login can subscribe to any topic in order for it to
respond to all feeders.

Currently there appears to be no way to filter what topics can be
published by a client.

### CODE REVIEW

**Comment by kruton** [comment]
@mbillow Feel free to blow this away and put in your own change. I just uploaded this in case it covers something you didn't do.

File: feeder/util/mqtt/authentication.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+local_username = "local_%s" % token_hex(8)
+local_password = token_hex(16)
+
+
+class PetnetAuthPlugin(BaseAuthPlugin):
+    username_regex = re.compile(r'^/pegasus:(?P<gateway_id>.*)$')
+
+    def __init__(self, context):
```
Comment (mbillow) [inline feeder/util/mqtt/authentication.py:16]: You shouldn’t need this constructor. I know they have it in the library code, but a child class will just inherit its parent constructor, so there is no need to make our own if all we are doing is calling super.

File: feeder/util/mqtt/authentication.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        super().__init__(context)
+
+    @asyncio.coroutine
```
Comment (mbillow) [inline feeder/util/mqtt/authentication.py:19]: This wrapper is also redundant (I think 🤔)... Using async def is just the >Python 3.4 syntactic sugar for an async coroutine.

File: feeder/util/mqtt/client.py [hunk 3]
```diff
@@ -155,7 +155,7 @@ async def send_cmd_schedule(
         )
 
     async def start(self):
-        await self.connect("mqtt://localhost:1883/")
+        await self.connect("mqtt://%s:%s@localhost:1883/" % (local_username, local_password))
```
Comment (mbillow) [inline feeder/util/mqtt/client.py:158]: You can just use a format string here: `f"mqtt://{local_username}:{local_password}@localhost:1883/`

I’ve always found them about 100x easier to grok when trying to understand code.

File: feeder/util/mqtt/topic.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+from hbmqtt.plugins.topic_checking import BaseTopicPlugin
+from feeder.util.mqtt.authentication import local_username
+
+
+class PetnetTopicPlugin(BaseTopicPlugin):
+    feeder_sub_topic_regex = re.compile(
+        r"krs/(api|cmd)/stg/(?P<gateway_id>.*)$")
+    username_regex = re.compile(r'^/pegasus:(?P<gateway_id>.*)$')
+
+    def __init__(self, context):
```
Comment (mbillow) [inline feeder/util/mqtt/topic.py:13]: Same comments here I left on the auth plugin.

**Review by mbillow — CHANGES_REQUESTED** [review]
@kruton This is really great. I think the only difference in our changes is that I was generating a secret key as well and had a separate table for storing API creds. I doubt we will ever need the secret keys since it is only used in their request signing process. 

Also, the extra credit of topic filtering is a great idea. You’ve earned yourself a gold star! 😂

File: feeder/util/mqtt/topic.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    @asyncio.coroutine
+    async def topic_filtering(self, *args, **kwargs):
+        filter_result = super().topic_filtering(*args, **kwargs)
+        if not filter_result:
+            return False
+
+        session = kwargs.get('session', None)
+        topic = kwargs.get('topic', None)
+        self.context.logger.debug('username: %s, topic: %s' %
```
Comment (mbillow) [inline feeder/util/mqtt/topic.py:24]: Oh, you should also be passing your logging variables directly to the logger call along with the format string. This lets the logging module handle the substitution.

E.g. `info(“some message %s with %s”, var1, var2)`

**Review by mbillow — APPROVED** [review]

**Pull request merged**